File: problems/2-medium/74-Search-A-2D-Matrix/solution.py
from typing import List

def searchMatrix(matrix: List[List[int]], target: int) -> bool:
    m = len(matrix)
    n = len(matrix[0])
    
    left = 0
    right = m * n - 1

    while left <= right:
        mid = (left + right) // 2

        row = mid // n
        col = mid % n

        if target == matrix[row][col]:
            return True

        elif target < matrix[row][col]:
            right = mid - 1
        else:
            left = mid + 1
        
    return False

File: problems/2-medium/74-Search-A-2D-Matrix/solution.java
public boolean searchMatrix(int[][] matrix, int target) {
    int left = 0;
    int right = matrix.length - 1;
    int index = 0;
    

    while (left <= right) {
        index = (right + left) / 2;
        if (matrix[index][0] > target) {
            right = index - 1;
        } else if (matrix[index][matrix[index].length - 1] < target) {
            left = index + 1;
        } else {
            break;
        }
    }
    
    left = 0;
    right = matrix[index].length - 1;
    int inner_index = 0;
    while (left <= right) {
        inner_index = (right + left) / 2;
        if (matrix[index][inner_index] > target) {
            right = inner_index - 1;
        } else if (matrix[index][inner_index] < target) {
            left = inner_index + 1;
        } else { 
            return true;
        } 
    }      
    return false;
}

